### PR TITLE
lowering: Disallow splatting in non-final default value

### PR DESCRIPTION
**Pull request by Keno**
Pop quiz: Do you know what the following will do?
```
julia> function g1(a=(1,2)..., b...=3)
    b
end

julia> g1()

julia> function g2(a=(1,2)..., b=3, c=4)
    (b, c)
end

julia> g2()

julia> function g3(a=(1,2)..., b=3, c...=4)
    (b, c)
end

julia> g3()

julia> g3(1)
```

I don't either and I don't think it's particularly well defined. Splatting a default argument makes sense on the last argument, which can be a vararg, and it is desirable to be able to specify the default for the whole varargs tuple at once (although arguably that should just be the non-`...` behavior, but that'd be too breaking a change). Ref #50518. However, for other arguments, there isn't really a sensible semantic meaning. This PR disallows this in lowering. This is technically a minor change, but I doubt anybody is using this. Splatting in default values wasn't really ever supposed to work anyway, it just happened to fall out of our lowering.

### CODE REVIEW

**Comment by ericphanson** [comment]
Is the variant from the issue still allowed?
```julia
function f(xs...=["a", "b", "c"]...; debug=false)
           return xs[1]
end
```
Maybe worth a test?

**Comment by Keno** [comment]
Yes, it's allowed and was added as a test in #50559.

File: src/julia-syntax.scm [hunk 1]
```diff
@@ -662,7 +662,10 @@
                       (vals   (list-tail dfl n))
                       (absent (list-tail opt n)) ;; absent arguments
                       (body
-                       (if (any (lambda (defaultv)
+                       (if (any (lambda (val) (and (pair? val) (eq? (car val) '...))) (butlast vals))
```
Comment (JeffBezanson) [inline src/julia-syntax.scm:665]:
```suggestion
                       (if (any vararg? (butlast vals))
```

File: src/julia-syntax.scm [hunk 2]
```diff
@@ -662,7 +662,10 @@
...
+                         ;; Forbid splat in all but the final default value
+                         (error "Splatting in non-final positional default")
```
Comment (JeffBezanson) [inline src/julia-syntax.scm:667]:
```suggestion
                         (error "invalid \"...\" in non-final positional argument default value")
```

**Comment by maleadt** [comment]
> This is technically a minor change, but I doubt anybody is using this.

Let's see if Nanosoldier finds any packages:

@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/50563#issuecomment-1640068497) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/69d7385_vs_7141e73/report.html) is available.

**Pull request merged**